### PR TITLE
Optimized lock for per-pool locking

### LINKED ISSUE #2724: Optimize lock for per-pool locking
We've introduced multithreading into stratisd but have no yet managed to optimize a lot for per-pool locking.

### PR DESCRIPTION
**Pull request by jbaublitz**
Closes #2724
Supersedes #2725

### CODE REVIEW

**Comment by jbaublitz** [comment]
TODO: Fix occasional deadlock in devicemapper event handling by handling each pool event in its own thread.

File: src/dbus_api/udev.rs [hunk 1]
```diff
@@ -37,21 +41,43 @@ where
...
+        let mut events = Vec::new();
+        events.push(self.receiver.recv().await.ok_or_else(|| {
             StratisError::Msg("Channel from udev handler to D-Bus handler was shut".to_string())
-        })?;
-        let mut mutex_lock = self.dbus_context.engine.lock().await;
-        let optional_pool_info = mutex_lock.handle_event(&udev_event);
+        })?);
+        loop {
+            let recv = self.receiver.recv();
+            pin_mut!(recv);
```
Comment (mulkieran) [inline src/dbus_api/udev.rs:50]: Why does this need to be pinned?

**Comment by jbaublitz** [comment]
Deadlock appears to be fixed between the most recent commit and #2791. udev tests are no longer hanging on my local machine.

**Comment by jbaublitz** [comment]
@mulkieran @bgurney-rh Do either of you know why that `checks-with-ci-repo` is failing? It appears to be the same version of Rust and Fedora, but it's getting compilation errors that almost look like an old version of Rust before `OwnedRwLockReadGuard`, etc. were added.

**Comment by jbaublitz** [comment]
It appears that stratis-cli checks don't start udev which is why we're seeing these failures.

**Comment by jbaublitz** [comment]
> @mulkieran @bgurney-rh Do either of you know why that `checks-with-ci-repo` is failing? It appears to be the same version of Rust and Fedora, but it's getting compilation errors that almost look like an old version of Rust before `OwnedRwLockReadGuard`, etc. were added.

@mulkieran After looking into this further, it appears that your test is working as intended! `OwnedRwLockReadGuard` was added in tokio 1.5.

**Comment by jbaublitz** [comment]
@mulkieran I'll rebase this after #2796 and I think the tests should pass.

**Comment by jbaublitz** [comment]
@mulkieran Please see the most recent commit before reviewing. The stratis-cli tests actually use the sim engine so udev is actually not needed at all. If you're satisfied with the solution I put in place, feel free to review this PR and merge #2803. Otherwise, we should discuss how to better approach the problem. Does that sound good?

**Comment by mulkieran** [comment]
@jbaublitz Plz rebase, once more.

**Pull request closed without merging**